app/services/RestaurantService.java

```diff
@@ -1,157 +1,208 @@
-		}
-
-		criteria.addOrder(Order.asc("name"));
+	if (restaurantFilter.sortBy.equals("price"))
+	{
+	    criteria.addOrder(Order.desc("priceRange"));
+	}
 
-		List<Restaurant> restaurants = criteria.list();
+	criteria.addOrder(Order.asc("name"));
 
-		switch (restaurantFilter.sortBy) {
-			case "rating":
-				restaurants.sort((o1, o2) -> o2.getAverageRating().compareTo(o1.getAverageRating()));
-				break;
-		}
+	List<Restaurant> restaurants = criteria.list();
 
-		return PaginationAdapter.createOutput()
-				.setPageNumber(restaurantFilter.pageNumber)
-				.setPageSize(restaurantFilter.pageSize)
-				.setModel(restaurants)
-				.setNumberOfPages(numberOfPages);
+	switch (restaurantFilter.sortBy)
+	{
+	case "rating":
+	    restaurants.sort((o1, o2) -> o2.getAverageRating().compareTo(o1.getAverageRating()));
+	    break;
 	}
 
-	/**
-	 * Gets restaurant with id.
-	 *
-	 * @param id the id
-	 * @return the restaurant with id
-	 */
-	public Restaurant getRestaurantWithId(final UUID id) {
-		return (Restaurant) getSession().createCriteria(Restaurant.class)
-				.add(Restrictions.eq("id", id))
-				.uniqueResult();
+	return PaginationAdapter.createOutput().setPageNumber(restaurantFilter.pageNumber)
+		.setPageSize(restaurantFilter.pageSize).setModel(restaurants).setNumberOfPages(numberOfPages);
+    }
+
+    /**
+     * Gets restaurant with id.
+     *
+     * @param id
+     *            the id
+     * @return the restaurant with id
+     */
+    public Restaurant getRestaurantWithId(final UUID id)
+    {
+	return (Restaurant) getSession().createCriteria(Restaurant.class).add(Restrictions.eq("id", id)).uniqueResult();
+    }
+
+    /**
+     * Gets nearby restaurants.
+     *
+     * @param latitude
+     *            the latitude
+     * @param longitude
+     *            the longitude
+     * @return the nearby restaurants
+     */
+    @SuppressWarnings("unchecked")
+    public List<Restaurant> getNearbyRestaurants(final Float latitude, final Float longitude)
+    {
+	return getSession()
+		.createSQLQuery(
+			"SELECT * FROM restaurant WHERE restaurant.longitude <> 0 AND restaurant.latitude <> 0 ORDER BY ST_Distance(ST_GeomFromText('POINT(' || restaurant.longitude || ' ' || restaurant.latitude || ')' ,4326), ST_GeomFromText('POINT(' || :longitude || ' ' || :latitude || ')',4326)) ASC LIMIT 3")
+		.addEntity(Restaurant.class).setParameter("longitude", longitude).setParameter("latitude", latitude)
+		.list();
+    }
+
+    /**
+     * Gets popular restaurants.
+     *
+     * @return the popular restaurants
+     */
+    @SuppressWarnings("unchecked")
+    public List<Restaurant> getPopularRestaurants(UUID userId)
+    {
+
+	UserHabitsBean userHabits = getUserHabits(userId);
+      
+	List<Restaurant> restaurants = (List<Restaurant>) getSession().createCriteria(Restaurant.class)
+		.add(Restrictions.eq("city.id", userHabits.getCity()))
+		.add(Restrictions.eq("priceRange", userHabits.getUserPriceRange())).list();
```

Comment:
This lists all Restaurants which are in the same city as the user and in the specific price range? If those are the only restrictions, this can not work. They are not specific enough to prevent problems. Assume that it is not Sarajevo, but New York, or Tokyo. What happens then? 100k restaurants will be loaded to memory, and probably crashing the entire JVM.  Pagination has to be used.